### DISCUSSION TITLE
Timeout in propose()

### DISCUSSION THREAD

**Discussion by rnz1234** [yasserfarouk/scml#55] [one-shot-qa]
Hi, 
I see a very strange behavior when running parallel tournament.
I see that in some occasions when long initialization is happening during propose (the first time we see an opponent, we need to init some database used for learning), I see that the propose() method doesn't end within the same agent object (I observe when function is entered and ends per agent object) - before it end the on_negotiation_failure() callback is called.
Why does this happen ?
is there a timeout ?
is it on the real tournament as well ? if not how do I cancel this in my env ?
doesn't it make the entire engine non-deterministic ?

Thanks

**Comment by yasserfarouk** [comment]
That is expected behavior. There is a timeout of 10seconds for responding. `propose()` is called on a dedicated thread that is ignored after the timeout leading to the failure (Unfortunately it is not killed immediately because of a technical issue). You can control the value of the time-limit by passing `neg_step_time_limit` and `neg_time_out` explicitly. Passing `float("inf")` will disable both.

In all tournaments including the official competition, a timeout will be used. We will try to make it as large as we can given the limitations on computation power we have but I expect that it will not exceed 20sec per response in any case. 

I am not sure what you mean by the game being non-deterministic but having a timeout condition is necessary given possibilities of infinite loops and extremely long processing in agent code that we do not control. If what is meant is that you cannot assume a specific ordering on method calls, then yes you cannot in general but there are some useful rules:

- Any call to `propse()`, `respond()` can assume that all previous such calls have ended successfully (otherwise the negotiation would have timed out).
- `on_negotiation_success/failure()` **CANNOT** assume that any calls to `propose()` or `respond()` were completed as a timeout may have happened in the first such call (as in your case).
